cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll

```diff
@@ -10,29 +10,39 @@ private import semmle.code.cpp.ir.implementation.internal.OperandTag
 private class IntValue = Ints::IntValue;
 
 private predicate hasResultMemoryAccess(
-  Instruction instr, IRVariable var, Type type, IntValue startBitOffset, IntValue endBitOffset
+  Instruction instr, IRVariable var, IRType type, Language::LanguageType languageType,
+  IntValue startBitOffset, IntValue endBitOffset
 ) {
   resultPointsTo(instr.getResultAddress(), var, startBitOffset) and
-  type = instr.getResultType() and
-  if exists(instr.getResultSize())
-  then endBitOffset = Ints::add(startBitOffset, Ints::mul(instr.getResultSize(), 8))
+  languageType = instr.getResultLanguageType() and
+  type = languageType.getIRType() and
+  if exists(type.getByteSize())
+  then endBitOffset = Ints::add(startBitOffset, Ints::mul(type.getByteSize(), 8))
   else endBitOffset = Ints::unknown()
 }
 
 private predicate hasOperandMemoryAccess(
-  MemoryOperand operand, IRVariable var, Type type, IntValue startBitOffset, IntValue endBitOffset
+  MemoryOperand operand, IRVariable var, IRType type, Language::LanguageType languageType,
+  IntValue startBitOffset, IntValue endBitOffset
 ) {
   resultPointsTo(operand.getAddressOperand().getAnyDef(), var, startBitOffset) and
-  type = operand.getType() and
-  if exists(operand.getSize())
-  then endBitOffset = Ints::add(startBitOffset, Ints::mul(operand.getSize(), 8))
+  languageType = operand.getLanguageType() and
+  type = languageType.getIRType() and
+  if exists(type.getByteSize())
+  then endBitOffset = Ints::add(startBitOffset, Ints::mul(type.getByteSize(), 8))
   else endBitOffset = Ints::unknown()
 }
 
 private newtype TMemoryLocation =
-  TVariableMemoryLocation(IRVariable var, Type type, IntValue startBitOffset, IntValue endBitOffset) {
-    hasResultMemoryAccess(_, var, type, startBitOffset, endBitOffset) or
-    hasOperandMemoryAccess(_, var, type, startBitOffset, endBitOffset)
+  TVariableMemoryLocation(
+    IRVariable var, IRType type, Language::LanguageType languageType, IntValue startBitOffset,
+    IntValue endBitOffset
+  ) {
+    (
+      hasResultMemoryAccess(_, var, type, _, startBitOffset, endBitOffset) or
+      hasOperandMemoryAccess(_, var, type, _, startBitOffset, endBitOffset)
+    ) and
+    languageType = type.getCanonicalLanguageType()
```

Comment:
The "canonical language type" concept seems like an artificial workaround, but I'm not sure what it's a workaround for. Is it just to make the `AliasedSSA` changes minimal, or do you intend for it to remain for a long time?

Reply:
It is a workaround, but I don't have a long-term solution. The issue is that I'd like to ensure that every `Instruction` has a result for `getResultLanguageType()`. Most instructions just have their type determined during raw IR construction, where we can use the type from the AST. `Phi` and `Chi` instructions, though, get inserted during SSA construction. For a `Chi` instruction, the result type is just the type of the virtual variable. This is either the type of the actual `IRVariable`, or, in the case of the unknown (aliased) virtual variable, it's `UnknownType`. For `Phi` instructions, though, we have to infer the type from the multiple memory accesses to that memory location. They'll all have the same `IRType`, but there's no guarantee that they'll have the same `LanguageType`. For example:
```
typedef int MYINT;
MYINT x;
if (b) {
  x = 5;  // x1
}
else {
  int* px = &x;
  *px = 7;  // x2
}
// x3 = Phi(x1, x2)
return x;  // x3
```
The store in the `then` branch is of language type `MYINT`, but the store in the `else` branch is of type `int`. What should the result type of the `Phi` instruction be? Either we need to consider the types of all of the memory accesses and deterministically pick the "best" one, or we need a way to pick a single canonical `LanguageType` from a given `IRType`.

Reply:
For phi nodes, perhaps we could attempt to deterministically pick a "best" type and then fall back to `UnknownType` if we can't pick a unique best type. We could start out with a very simple implementation and only extend it if there's particular need.

Reply:
`Phi` nodes expect the result type to be equivalent to each operand type, so I don’t like falling back to `UnknownType`. I think I could move `getCanonicalLanguageType` onto `LanguageType`. The implementation would be about the same complexity, but the interface might be cleaner.

Reply:
What breaks if an instruction (an ambiguous phi instruction) has no result for `getResultLanguageType`?

Reply:
Right now, the implementation of `getResultIRType()` just calls `getResultLanguageType().getIRType()`, so we wouldn't be able to get an `IRType` for the result either.

Reply:
Let's not delay the PR further over this implementation wart. The exposed API seems fine.